### PR TITLE
Rust: Make new(), run() and show() report errors from the backend

### LINKED ISSUE #2198: Should run(), show() and hide() return an error?
Should the show, hide and run from the component return an error in rust.

Currently, we get panics within the backend when there is an error to, for example, connect to the display or get an opengl context.
Should we instead report an error through the show, hide, and run, on the slint::Window  and the component,  and on the `slint::run_event_loop`? 

In case of an error for which we currently panic, we should exit the event loop and return from run_event_loop with an error.

### PR DESCRIPTION
**Pull request by ogoffart**
Fixes #2198

### CODE REVIEW

**Comment by ogoffart** [comment]
Note: I didn't make hide() return an error.

File: internal/backends/qt/lib.rs [hunk 1]
```diff
@@ -127,12 +128,14 @@ pub const HAS_NATIVE_STYLE: bool = cfg!(not(no_qt));
 
 pub struct Backend;
 impl i_slint_core::platform::Platform for Backend {
-    fn create_window_adapter(&self) -> Rc<dyn i_slint_core::window::WindowAdapter> {
+    fn create_window_adapter(
+        &self,
+    ) -> Result<Rc<dyn i_slint_core::window::WindowAdapter>, PlatformError> {
         #[cfg(no_qt)]
-        panic!("The Qt backend needs Qt");
+        return Err("Qt platform being used despite Slint is compiled without Qt support".into());
```
Comment (tronical) [inline internal/backends/qt/lib.rs:135]:
```suggestion
        return Err("Qt platform requested but Slint is compiled without Qt support".into());
```

File: internal/backends/qt/lib.rs [hunk 2]
```diff
@@ -158,7 +161,10 @@ impl i_slint_core::platform::Platform for Backend {
                 ensure_initialized(true);
                 qApp->exec();
             } }
+            return Ok(());
         };
+        #[cfg(no_qt)]
+        Err("Qt platform being used despite Slint is compiled without Qt support".into())
```
Comment (tronical) [inline internal/backends/qt/lib.rs:167]:
```suggestion
        Err("Qt platform requested but Slint is compiled without Qt support".into())
```

File: internal/core/api.rs [hunk 3]
```diff
@@ -741,6 +742,64 @@ pub enum EventLoopError {
...
+/// Either it couldn't be initialized, or some problem happened with the display server
+///
+/// This can be constructed from a `String`:
+/// ```rust
+/// use slint::platform::PlatformError;
+/// PlatformError::from(format!("Could not load resource {}", 1234));
+/// ```
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub enum PlatformError {
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/core/api.rs:756]: I like it!

File: internal/core/api.rs [hunk 4]
```diff
@@ -741,6 +742,64 @@ pub enum EventLoopError {
...
+///
+/// This can be constructed from a `String`:
+/// ```rust
+/// use slint::platform::PlatformError;
+/// PlatformError::from(format!("Could not load resource {}", 1234));
+/// ```
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum PlatformError {
+    /// There was no default platform selected, and no platform was initialized
```
Comment (tronical) [inline internal/core/api.rs:757]:
```suggestion
    /// No default platform was selected, or no platform could be initialized.
```

File: internal/core/api.rs [hunk 5]
```diff
@@ -741,6 +742,64 @@ pub enum EventLoopError {
     NoEventLoopProvider,
 }
 
+/// The platform encountered a fatal error.
+///
+/// Either it couldn't be initialized, or some problem happened with the display server
```
Comment (tronical) [inline internal/core/api.rs:747]:
```suggestion
/// This error typically indicates an issue with initialization or connecting to the windowing system.
```

File: internal/core/api.rs [hunk 6]
```diff
@@ -741,6 +742,64 @@ pub enum EventLoopError {
...
+/// PlatformError::from(format!("Could not load resource {}", 1234));
+/// ```
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum PlatformError {
+    /// There was no default platform selected, and no platform was initialized
+    ///
+    /// Either a platform must be selected trough the `backend-*` cargo features flags,
+    /// or [`platform::set_platform()`](crate::platform::set_platform) must be called
+    /// before running the event loop
```
Comment (tronical) [inline internal/core/api.rs:761]:
```suggestion
    /// If you encounter this error, make sure to either selected trough the `backend-*` cargo features flags,
    /// or call [`platform::set_platform()`](crate::platform::set_platform)
    /// before running the event loop
```

File: internal/core/api.rs [hunk 7]
```diff
@@ -741,6 +742,64 @@ pub enum EventLoopError {
...
+    NoPlatform,
+    /// The Slint Platform does not provide an event loop.
+    ///
+    /// The [`Platform::run_event_loop`](crate::platform::Platform::run_event_loop)
+    /// is not implemented for the current platform
```
Comment (tronical) [inline internal/core/api.rs:766]:
```suggestion
    /// is not implemented for the current platform.
```

File: internal/core/api.rs [hunk 8]
```diff
@@ -741,6 +742,64 @@ pub enum EventLoopError {
...
+    NoEventLoopProvider,
+
+    /// There is already a platform set from another thread another thread
```
Comment (tronical) [inline internal/core/api.rs:769]:
```suggestion
    /// There is already a platform set from another thread.
```

File: internal/core/api.rs [hunk 9]
```diff
@@ -741,6 +742,64 @@ pub enum EventLoopError {
...
+
+    /// Another platform-specific error occurred
+    Other(String),
+}
+
+impl core::fmt::Display for PlatformError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            PlatformError::NoPlatform => f.write_str(
+                "There was no default Slint platform selected, and no Slint platform was initialized",
```
Comment (tronical) [inline internal/core/api.rs:780]: Just formatting to use "XXX was selected, XXX was initialized" in both parts of the sentence.

```suggestion
                "No default Slint platform was selected, and no Slint platform was initialized",
```

**Review by tronical — APPROVED** [review]

**Pull request merged**